### DISCUSSION TITLE
React Hook Experience

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#73] [프론트 기술 질문 공간]
리액트 훅에 대해 얼마나 알고 계신가요?
사용해본 훅이 있다면 설명해주세요.

**Comment by michoball** [comment]
훅은 리액트에서 클래스 컴포넌트가 아닌 함수형  컴포넌트를 사용하는 쪽으로 방향이 바뀌며 등장했습니다. 
함수형 컴포넌트에서 데이터의 상태나 side effect를 분리해 다루기 위해 사용됩니다.  

대표적인 훅으로는 useState, useEffect, useRef, useCallback, useMemo 등이 있습니다. 

리액트로 프로그래밍을 할 때는 거의 모든 경우 위에 서술한 훅을 사용합니다. 

useState: 컴포넌트 내 데이터의 상태를 다루기위한 훅
useEffect: side effect 를 다루기 위한 훅
useRef: 컴포넌트에서 렌더링 되지 않아도되는 값을 참조하기위한 훅
useCallback: 컴포넌트 리렌더링 시 함수를 캐시해서 특정 조건에만 함수가 변하도록하는 훅
useMemo: useCallback과 같은이유로 값을 캐싱하는 훅

이외에 props drilling을 해결하고자 useContext를 사용하거나 여러가지 훅들을 간편히 사용하기위해 커스텀 훅을 만들어 사용하기도 합니다.

**Comment by WooGie911** [comment]
Hook은 Class없이 React 기능들을 사용하는 방법을 제시합니다.
React 에서의 Class 사용을 위해서는 JavaScript의 this 키워드가 어떻게 작동하는지 알아야만 합니다. 
JavaScript의 this키워드는 대부분의 다른 언어에서와는 다르게 작동함으로 사용자에게 큰 혼란을 주었으며, 코드의 재사용성과 구성을 매우 어렵게 만들고는 했습니다. 
이와 같은 어려움을 해결하기 위해, 16.8 버전부터 'Hook'이 도입되었습니다. 

기존의 클래스형 컴포넌트의 단점 :
- 상태(State) 로직 재사용 어려움
- 코드가 복잡해짐
- 관련 없는 로직들이 함께 섞여 있어 이해가 힘듬

Hook 종류
- [useState]
- [useEffect]
- [useRef]
- [useContext]
- [useMemo]
- [useCallback]
- [useDispatch]
- [useSelector]

**Comment by javascriptbiggosoo** [comment]
리액트 컴포넌트에서 사용하는 use로 시작하는 모든 함수가 훅입니다.
리액트 컴포넌트, 커스텀 훅이 아닌 고에서는 사용할 수 없습니다.
그 안에서도 최상위 레벨에서만 사용가능하며 중첩된 블록에서는 사용할 수 없습니다.
